framework_agreement/model/portfolio.py

```diff
@@ -35,15 +42,212 @@ def get_from_supplier(self, supplier):
+            })
+
+    @api.multi
+    def get_line_for_product(self, product, quantity=0):
+        self.ensure_one()
+        for product_line in self.line_ids:
+            if (product_line.product_id == product and
+                    product_line.quantity >= quantity):
+                return product_line
+        return False
+
+    @api.multi
+    def is_suitable_for(self, date, product, quantity=0):
+        # we prefer doing a rough check and let the user make their choices
+        # instead of risking to limit things too much for a small discrepancy.
+        # Think of an agreement that would be good a few days later, or with a
+        # slightly different incoterm.
+        return self.state in ('running', 'future')
+
+    @api.multi
+    def is_valid_at_date(self, proposed_date):
+        self.ensure_one()
+        start = datetime.strptime(self.start_date, DATE_FORMAT)
+        is_valid = start < proposed_date
+
+        if self.end_date:
+            end = datetime.strptime(self.end_date, DATE_FORMAT)
+            is_valid = is_valid and proposed_date < end
+
+        return is_valid
+
+
+class AgreementProductLine(models.Model):
+    _name = 'agreement.product.line'
+
+    portfolio_id = fields.Many2one('framework.agreement.portfolio',
+                                   required=True)
+    product_id = fields.Many2one('product.product', 'Product', required=True)
+    quantity = fields.Integer('Negociated quantity', required=True)
+    available_quantity = fields.Integer(
+        compute='_compute_available_qty',
+        string='Available quantity',
+        store=True,
+        # this is stored because we do searches on it. However, the method
+        # might be called more often than necessary (specifically, when
+        # computing the state of the portfolio).
+    )
+
+    @api.depends(
+        'quantity',
+        'product_id',
+        'portfolio_id',
+        'portfolio_id.purchase_ids',
+        'portfolio_id.purchase_ids.portfolio_id',
+        'portfolio_id.purchase_ids.state',
+        'portfolio_id.purchase_ids.order_line.product_qty'
+    )
+    @api.one
+    def _compute_available_qty(self):
+        """Compute available qty based on confirmed PO lines."""
+        if isinstance(self.portfolio_id.id, models.NewId):
+            return
+
+        sql = """
+            SELECT SUM(po_line.product_qty)
+            FROM purchase_order_line AS po_line
+            LEFT JOIN purchase_order AS po ON po_line.order_id = po.id
+            WHERE po.portfolio_id = %(portfolio)s
+            AND po.state IN %(states)s
+            AND po.company_id = %(company)s
+            AND po_line.product_id = %(product)s
+        """
+        self.env.cr.execute(sql, {
+            'portfolio': self.portfolio_id.id,
+            'states': AGR_PO_STATE,
+            'company': self.portfolio_id._company_get().id,
+            'product': self.product_id.id,
+        })
+        used_amount = self.env.cr.fetchone()[0] or 0
```

Comment:
You do not take in account the use case where you SQL do not have a match and is None